catris/games/traditional.py

```diff
@@ -99,29 +99,33 @@ def find_and_then_wipe_full_lines(self) -> Iterator[set[Square]]:
 
         self.finish_wiping_full_lines()
 
-    def add_player(self, name: str, color: int) -> Player:
-        self.players.append(
-            Player(
-                name,
-                color,
-                up_x=0,
-                up_y=-1,
-                moving_block_start_x=123,  # changed below
-                moving_block_start_y=-1,
-            )
-        )
-
-        # width per player might have changed, adjust spawning locations
+    def _update_spawn_places_and_landed_coords(self) -> None:
         w = self._get_width_per_player()
         for i, player in enumerate(self.players):
             player.moving_block_start_x = (i * w) + (w // 2)
 
-        for y in range(self.HEIGHT):
-            for x in range(self._get_width()):
-                self.valid_landed_coordinates.add((x, y))
+        self.valid_landed_coordinates = {
+            (x, y) for x in range(self._get_width()) for y in range(self.HEIGHT)
+        }
 
+    def add_player(self, name: str, color: int) -> Player:
+        player = Player(
+            name,
+            color,
+            up_x=0,
+            up_y=-1,
+            moving_block_start_x=123,  # changed soon
+            moving_block_start_y=-1,
+        )
+        self.players.append(player)
+        self._update_spawn_places_and_landed_coords()
+        self.new_block(player)
         return player
 
+    def remove_player(self, player: Player) -> None:
+        self.players.remove(player)
+        self._update_spawn_places_and_landed_coords()
```

Comment:
- Should move blocks so that they don't end up outside allowed landing places
- What happens if player has wait counter going?